File: backend/app/schemas/master.py
from datetime import datetime
from uuid import UUID
from typing import Any, Generic, Optional, TypeVar

from pydantic import BaseModel, ConfigDict, Field


T = TypeVar("T")


class UserBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    first_name: str
    last_name: Optional[str] = None
    avatar_url: Optional[str] = None
    city: Optional[str] = None


class CategoryBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    slug: str


class ServiceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    master_id: UUID
    name: str
    description: Optional[str] = None
    price: float
    duration_minutes: int
    is_active: bool
    created_at: datetime


class ServiceCreate(BaseModel):
    name: str
    description: Optional[str] = None
    price: float = Field(..., gt=0)
    duration_minutes: int = Field(..., gt=0)


class ServiceUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    price: Optional[float] = None
    duration_minutes: Optional[int] = None
    is_active: Optional[bool] = None


class PortfolioImageResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    image_url: str
    sort_order: int


class MasterProfileCreate(BaseModel):
    description: Optional[str] = None
    category_id: UUID
    district: Optional[str] = None
    work_hours: Optional[dict[str, Any]] = None


class MasterProfileUpdate(BaseModel):
    description: Optional[str] = None
    category_id: Optional[UUID] = None
    district: Optional[str] = None
    work_hours: Optional[dict[str, Any]] = None
    is_available: Optional[bool] = None


class MasterListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user: UserBrief
    category: CategoryBrief
    description: Optional[str] = None
    district: Optional[str] = None
    rating_avg: float
    rating_count: int
    verification_status: str
    is_available: bool
    price_from: Optional[float] = None
    services_count: int


class MasterDetail(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user: UserBrief
    category: CategoryBrief
    description: Optional[str] = None
    district: Optional[str] = None
    rating_avg: float
    rating_count: int
    verification_status: str
    is_available: bool
    price_from: Optional[float] = None
    services_count: int
    work_hours: Optional[dict[str, Any]] = None
    services: list[ServiceResponse] = []
    portfolio: list[PortfolioImageResponse] = []
    reviews_preview: list[Any] = []


class PaginatedResponse(BaseModel, Generic[T]):
    items: list[T]
    total: int
    page: int
    page_size: int
    pages: int

File: mobile/lib/shared/models/master.dart
import 'category.dart';
import 'service.dart';
import 'review.dart';

class MasterUser {
  final String id;
  final String firstName;
  final String? lastName;
  final String? avatarUrl;
  final String? city;

  MasterUser({
    required this.id,
    required this.firstName,
    this.lastName,
    this.avatarUrl,
    this.city,
  });

  String get fullName => [firstName, lastName].where((e) => e != null && e.isNotEmpty).join(' ');

  factory MasterUser.fromJson(Map<String, dynamic> json) {
    return MasterUser(
      id: json['id'] as String,
      firstName: json['first_name'] as String,
      lastName: json['last_name'] as String?,
      avatarUrl: json['avatar_url'] as String?,
      city: json['city'] as String?,
    );
  }
}

class Master {
  final String id;
  final MasterUser user;
  final Category? category;
  final String? description;
  final String? district;
  final double ratingAvg;
  final int ratingCount;
  final String verificationStatus;
  final bool isAvailable;
  final double? priceFrom;
  final int? servicesCount;
  final Map<String, dynamic>? workHours;
  final List<Service>? services;
  final List<PortfolioImage>? portfolio;
  final List<Review>? reviewsPreview;

  Master({
    required this.id,
    required this.user,
    this.category,
    this.description,
    this.district,
    this.ratingAvg = 0,
    this.ratingCount = 0,
    this.verificationStatus = 'pending',
    this.isAvailable = true,
    this.priceFrom,
    this.servicesCount,
    this.workHours,
    this.services,
    this.portfolio,
    this.reviewsPreview,
  });

  bool get isVerified => verificationStatus == 'verified';

  factory Master.fromJson(Map<String, dynamic> json) {
    return Master(
      id: json['id'] as String,
      user: MasterUser.fromJson(json['user'] as Map<String, dynamic>),
      category: json['category'] != null
          ? Category.fromJson(json['category'] as Map<String, dynamic>)
          : null,
      description: json['description'] as String?,
      district: json['district'] as String?,
      ratingAvg: (json['rating_avg'] as num?)?.toDouble() ?? 0,
      ratingCount: json['rating_count'] as int? ?? 0,
      verificationStatus: json['verification_status'] as String? ?? 'pending',
      isAvailable: json['is_available'] as bool? ?? true,
      priceFrom: (json['price_from'] as num?)?.toDouble(),
      servicesCount: json['services_count'] as int?,
      workHours: json['work_hours'] as Map<String, dynamic>?,
      services: (json['services'] as List<dynamic>?)
          ?.map((e) => Service.fromJson(e as Map<String, dynamic>))
          .toList(),
      portfolio: (json['portfolio'] as List<dynamic>?)
          ?.map((e) => PortfolioImage.fromJson(e as Map<String, dynamic>))
          .toList(),
      reviewsPreview: (json['reviews_preview'] as List<dynamic>?)
          ?.map((e) => Review.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

class PortfolioImage {
  final String id;
  final String imageUrl;
  final int sortOrder;

  PortfolioImage({
    required this.id,
    required this.imageUrl,
    this.sortOrder = 0,
  });

  factory PortfolioImage.fromJson(Map<String, dynamic> json) {
    return PortfolioImage(
      id: json['id'] as String,
      imageUrl: json['image_url'] as String,
      sortOrder: json['sort_order'] as int? ?? 0,
    );
  }
}
